lib/text/lzw.flow

```diff
@@ -0,0 +1,73 @@
+import string;
+import ds/list;
+import ds/tree;
+import text/blueprint;
+
+export {
+	// LZW-compress a string
+	lzw_encode(s : string) -> [int];
+
+	// Decompress an LZW-encoded string
+	lzw_decode(data : [int]) -> string;
+}
+
+lzw_encode(s : string) -> [int] {
+	dict : ref Tree<[int], int> = ref makeTree();
+	out : ref List<int> = ref makeList();
+	phrase : ref [int]= ref [getCharCodeAt(s, 0)];
+	code = ref 65536;
+	fori(1, strlen(s)-1, \i -> {
+		currChar = getCharCodeAt(s, i);
+		newPrase = arrayPush(^phrase, currChar);
+		if (isSome(lookupTree(^dict, newPrase))) {
+			refArrayPush(phrase,  currChar);
+		} else {
+			out := Cons(
+				if (length(^phrase) > 1) {
+					lookupTreeDef(^dict, ^phrase, 0);
+				} else {
+					^phrase[0];
+				},
+				^out
+			);
+			dict := setTree(^dict, newPrase, ^code);
+			code := ^code + 1;
+			phrase := [currChar];
+		}
+	});
+
+	out := Cons(
+		if (length(^phrase) > 1) {
+			lookupTreeDef(^dict, ^phrase, 0);
+		} else {
+			^phrase[0];
+		},
+		^out
+	);
+
+	list2array(^out)
+}
+
+// Decompress an LZW-encoded string
+lzw_decode(data : [int]) -> string {
+	dict : ref Tree<int, [int]> = ref makeTree();
+	currChar : ref int = ref data[0];
+	out : ref List<[int]> = ref makeList1([^currChar]);
+	oldPhrase : ref [int] = ref [(^currChar)];
+	code = ref 65536;
+	fori (1, length(data)-1, \i -> {
+		currCode = data[i];
+		phrase = if (currCode < 65536) {
+			[data[i]];
+		} else {
+		   lookupTreeDef(^dict, currCode, arrayPush(^oldPhrase, ^currChar));
```

Comment:
add lookupTreeDefLazy based on eitherFn to call here:
`
lookupTreeDefLazy(^dict, currCode, \ -> arrayPush(^oldPhrase, ^currChar));
`